File: examples/python/subscriber.py
import json
import os
import urllib.request

GATEWAY_URL = os.getenv("GATEWAY_URL", "http://127.0.0.1:8787")
TOPIC = os.getenv("TOPIC", "alerts")
SUBSCRIBER_ID = os.getenv("SUBSCRIBER_ID", "py-sub-1")

request = urllib.request.Request(
    f"{GATEWAY_URL}/v1/stream",
    method="GET",
    headers={
        "x-sse-topic": TOPIC,
        "x-sse-id": SUBSCRIBER_ID,
    },
)

with urllib.request.urlopen(request, timeout=15) as response:
    event_id = None
    event_data: list[str] = []

    while True:
        raw_line = response.readline()
        if not raw_line:
            break

        line = raw_line.decode("utf-8").strip()
        if not line:
            if event_id and event_data:
                print(json.dumps({"id": event_id, "data": "\n".join(event_data)}))
                break
            event_id = None
            event_data.clear()
            continue

        if line.startswith("id: "):
            event_id = line[4:]
        elif line.startswith("data: "):
            event_data.append(line[6:])

File: examples/rust/src/bin/subscriber.rs
use std::env;
use std::io::{BufRead, BufReader};

fn getenv(key: &str, fallback: &str) -> String {
    env::var(key).unwrap_or_else(|_| fallback.to_string())
}

fn main() {
    let gateway_url = getenv("GATEWAY_URL", "http://127.0.0.1:8787");
    let topic = getenv("TOPIC", "alerts");
    let subscriber_id = getenv("SUBSCRIBER_ID", "rust-sub-1");

    let client = reqwest::blocking::Client::new();
    let response = client
        .get(format!("{}/v1/stream", gateway_url))
        .header("x-sse-topic", topic)
        .header("x-sse-id", subscriber_id)
        .send()
        .expect("subscribe request failed");

    if response.status().as_u16() != 200 {
        panic!("subscribe failed: {}", response.status().as_u16());
    }

    let mut reader = BufReader::new(response);
    let mut line = String::new();
    let mut event_id: Option<String> = None;
    let mut data_lines: Vec<String> = Vec::new();

    loop {
        line.clear();
        let bytes = reader
            .read_line(&mut line)
            .expect("failed to read stream line");
        if bytes == 0 {
            break;
        }

        let trimmed = line.trim_end_matches(['\r', '\n']);
        if trimmed.is_empty() {
            if let Some(id) = &event_id {
                if !data_lines.is_empty() {
                    println!(
                        "{{\"id\":\"{}\",\"data\":\"{}\"}}",
                        id,
                        data_lines.join("\\n")
                    );
                    return;
                }
            }
            continue;
        }

        if let Some(rest) = trimmed.strip_prefix("id: ") {
            event_id = Some(rest.to_string());
            continue;
        }

        if let Some(rest) = trimmed.strip_prefix("data: ") {
            data_lines.push(rest.to_string());
        }
    }

    panic!("stream closed without receiving event");
}
